internal/api/attestationconfigapi/cli/client/reporter.go

```diff
@@ -0,0 +1,358 @@
+		obj, err := client.Fetch(ctx, s3Client, apiVersionObject{version: date + ".json", variant: variant, cached: true})
+		if err != nil {
+			return *new(T), "", fmt.Errorf("get object: %w", err)
+		}
+		obj.variant = variant // variant is not set by Fetch, set it manually
+
+		if minimalVersion == nil {
+			v := obj.getVersion().(T)
+			minimalVersion = &v
+			minimalDate = date
+			continue
+		}
+
+		// If the current minimal version has newer versions than the one we just fetched,
+		// update the minimal version to the older version.
+		if isInputNewerThanOtherVersion(variant, *minimalVersion, obj.getVersion()) {
+			v := obj.getVersion().(T)
+			minimalVersion = &v
+			minimalDate = date
+		}
+	}
+
+	return *minimalVersion, minimalDate, nil
+}
+
+func isInputNewerThanOtherVersion(variant variant.Variant, inputVersion, otherVersion any) bool {
+	var result bool
+	actionForVariant(variant,
+		func() {
+			input := inputVersion.(attestationconfigapi.TDXVersion)
+			other := otherVersion.(attestationconfigapi.TDXVersion)
+			result = isInputNewerThanOtherTDXVersion(input, other)
+		},
+		func() {
+			input := inputVersion.(attestationconfigapi.SEVSNPVersion)
+			other := otherVersion.(attestationconfigapi.SEVSNPVersion)
+			result = isInputNewerThanOtherSEVSNPVersion(input, other)
+		},
+	)
+	return result
+}
+
+type apiVersionObject struct {
+	version string          `json:"-"`
+	variant variant.Variant `json:"-"`
+	cached  bool            `json:"-"`
+	snp     attestationconfigapi.SEVSNPVersion
+	tdx     attestationconfigapi.TDXVersion
+}
+
+func (a apiVersionObject) MarshalJSON() ([]byte, error) {
+	var res []byte
+	var err error
+	actionForVariant(a.variant,
+		func() {
+			res, err = json.Marshal(a.tdx)
+		},
+		func() {
+			res, err = json.Marshal(a.snp)
+		},
+	)
+	return res, err
+}
+
+func (a *apiVersionObject) UnmarshalJSON(data []byte) error {
+	errTDX := json.Unmarshal(data, &a.tdx)
+	errSNP := json.Unmarshal(data, &a.snp)
+	if errTDX == nil || errSNP == nil {
+		return nil
+	}
+	return fmt.Errorf("trying to unmarshal data into both TDX and SNP versions: %w", errors.Join(errTDX, errSNP))
+}
+
+// JSONPath returns the path to the JSON file for the request to the config api.
+// This is the path to the cached version in the S3 bucket.
+func (a apiVersionObject) JSONPath() string {
+	if a.cached {
+		return path.Join(reportVersionDir(a.variant), a.version)
+	}
+	return path.Join(attestationconfigapi.AttestationURLPath, a.variant.String(), a.version)
+}
+
+// ValidateRequest validates the request.
+func (a apiVersionObject) ValidateRequest() error {
+	if !strings.HasSuffix(a.version, ".json") {
+		return fmt.Errorf("version has no .json suffix")
+	}
+	return nil
+}
+
+// Validate is a No-Op at the moment.
+func (a apiVersionObject) Validate() error {
+	return nil
+}
+
+func (a apiVersionObject) getVersion() any {
+	var res any
+	actionForVariant(a.variant,
+		func() {
+			res = a.tdx
+		},
+		func() {
+			res = a.snp
+		},
+	)
```

Comment:
Would be clearer with a switch statement.

Reply:
See my other comment.
Using the same `actionForVariant` function in all cases where we need to perform different actions depending on the variant allows us to more easily add new versions without having to perform many code changes.